python_bindings/test/apps/bilateral_grid_generator.py

```diff
@@ -0,0 +1,196 @@
+        blurx, blury, blurz = _funcs("blurx", "blury", "blurz")
+        blurz[x, y, z, c] = (
+            histogram[x, y, z - 2, c]
+            + histogram[x, y, z - 1, c] * 4
+            + histogram[x, y, z, c] * 6
+            + histogram[x, y, z + 1, c] * 4
+            + histogram[x, y, z + 2, c]
+        )
+        blurx[x, y, z, c] = (
+            blurz[x - 2, y, z, c]
+            + blurz[x - 1, y, z, c] * 4
+            + blurz[x, y, z, c] * 6
+            + blurz[x + 1, y, z, c] * 4
+            + blurz[x + 2, y, z, c]
+        )
+        blury[x, y, z, c] = (
+            blurx[x, y - 2, z, c]
+            + blurx[x, y - 1, z, c] * 4
+            + blurx[x, y, z, c] * 6
+            + blurx[x, y + 1, z, c] * 4
+            + blurx[x, y + 2, z, c]
+        )
+
+        # Take trilinear samples to compute the output
+        val = hl.clamp(clamped[x, y], 0.0, 1.0)
+        zv = val / g.r_sigma
+        zi = hl.i32(zv)
+        zf = zv - zi
+        xf = hl.f32(x % g.s_sigma) / g.s_sigma
+        yf = hl.f32(y % g.s_sigma) / g.s_sigma
+        xi = x / g.s_sigma
+        yi = y / g.s_sigma
+
+        interpolated = hl.Func("interpolated")
+        interpolated[x, y, c] = hl.lerp(
+            hl.lerp(
+                hl.lerp(blury[xi, yi, zi, c], blury[xi + 1, yi, zi, c], xf),
+                hl.lerp(blury[xi, yi + 1, zi, c], blury[xi + 1, yi + 1, zi, c], xf),
+                yf,
+            ),
+            hl.lerp(
+                hl.lerp(blury[xi, yi, zi + 1, c], blury[xi + 1, yi, zi + 1, c], xf),
+                hl.lerp(
+                    blury[xi, yi + 1, zi + 1, c], blury[xi + 1, yi + 1, zi + 1, c], xf
+                ),
+                yf,
+            ),
+            zf,
+        )
+
+        # Normalize
+        g.bilateral_grid[x, y] = interpolated[x, y, 0] / interpolated[x, y, 1]
+
+        # ESTIMATES
+        # (This can be useful in conjunction with RunGen and benchmarks as well
+        # as auto-schedule, so we do it in all cases.)
+        # Provide estimates on the input image
+        g.input_buf.set_estimates([(0, 1536), (0, 2560)])
+        # Provide estimates on the parameters
+        g.r_sigma.set_estimate(0.1)
+        # TODO: Compute estimates from the parameter values
+        histogram.set_estimate(z, -2, 16)
+        blurz.set_estimate(z, 0, 12)
+        blurx.set_estimate(z, 0, 12)
+        blury.set_estimate(z, 0, 12)
+        g.bilateral_grid.set_estimates([(0, 1536), (0, 2560)])
+
+        if g.using_autoscheduler():
+            # nothing
+            pass
+        elif g.target().has_gpu_feature():
+            # 0.50ms on an RTX 2060
+
+            xi, yi, zi = _vars("xi", "yi", "zi")
+
+            # Schedule blurz in 8x8 tiles. This is a tile in
+            # grid-space, which means it represents something like
+            # 64x64 pixels in the input (if s_sigma is 8).
+            (blurz.compute_root().reorder(c, z, x, y).gpu_tile(x, y, xi, yi, 8, 8))
```

Comment:
Is it worth it to use parens for the one-liner schedules, or is it just confusing?

Reply:
Not sure. I basically wrapped all of the multi-schedule bits in parens and let Black do its thing, and, well, this is what it does. Not sure I like it but apparently that's the way the Python ecosystem seems to be moving. (Google is moving towards `pyink` internally, which is basically a fork of Black with a handful of minor changes to minimize churn.)

Reply:
I decided to remove all the parens from the cases that fit on a single line.